### ISSUE TITLE
MultibodyPlant should be `final`

### ISSUE THREAD

**Issue by EricCousineau-TRI** [RobotLocomotion/drake#10310]
We have tests that have classes that inherit from `MultibodyPlant` (e.g. `AxiallySymmetricFreeBodyPlant`).

The API of MBP has enough complex behavior hidden that it seems like a bad idea (to me) to encourage inheriting from this class. I believe we should prevent this.

**Comment by sherm1** [comment]
That would be unfortunate IMO. Very useful to be able to say that something is-a MBPlant.

**Comment by jwnimmer-tri** [comment]
I agree with Eric -- I don't think the type checking that comes from `MySpecialPlant : public MultibodyPlant` is worth the headaches of making the class design open.  For me, `MultibodyPlant* my_special_plant` is plenty good enough documentation (we have several instances of this formulation already in the code).

If you really _do_ need to keep it non-final, then you need to look under [System Scalar Conversion](https://drake.mit.edu/doxygen_cxx/group__system__scalar__conversion.html) at **Systems not marked as final** and repair MultibodyPlant to actually be correctly subclass-able.

**Comment by sherm1** [comment]
OK, good point. I'm happy to have it be final for now until we have a pressing need. Nothing would stop us from changing it later.

**Comment by EricCousineau-TRI** [comment]
Alrighty, sounds good! I'll take a pass at sealing it off sometime next week.

**Comment by jwnimmer-tri** [comment]
For the record, we had a long discussion about this in #10414 -- also concluding to mark it `final`.
